lingpy/sequence/smoothing.py

```diff
@@ -0,0 +1,760 @@
+    as in the original one in C by Geoffrey Sampson (1995; 2000; 2005; 2008)
+    (https://www.grsampson.net/Resources.html), and in the one by
+    Loper, Bird et al. (2001-2018, NLTK Project)
+    (http://www.nltk.org/_modules/nltk/probability.html). Please note that
+    due to minor differences in implementation intended to guarantee non-zero
+    probabilities even in cases of expected underflow, as well as our
+    relience on scipy's libraries for speed and our way of handling
+    probabilities that are not computable when the assumptions of SGT are
+    not met, most results will not exactly match those of the 'gold standard'
+    of Gale and Sampson, even though the differences are never expected to
+    be significative and are equally distributed across the samples.
+
+    Parameters
+    ----------
+
+    freqdist : dict
+        Frequency distribution of samples (keys) and counts (values) from
+        which the probability distribution will be calculated.
+
+    p_value : float
+        The p-value for calculating the confidence interval of the empirical
+        Turing estimate, which guides the decision of using either the Turing
+        estimate "x" or the loglinear smoothed "y". Defaults to 0.05, as per
+        the reference implementation by Sampson, but consider that the authors,
+        both in their paper and in the code following suggestions credited to
+        private communication with Fan Yang, consider using a value of 0.1.
+
+    allow_fail : bool
+        A logic value informing if the function is allowed to fail, throwing
+        RuntimeWarning exceptions, if the essential assumptions on the
+        frequency distribution are not met, i.e., if the slope of the loglinear
+        regression is > -1.0 or if an unobserved count is reached before we are
+        able to cross the smoothing threshold. If set to False, the estimation
+        might result in an unreliable probability distribution; defaults to
+        True.
+
+    default_p0 : float
+        An optional value indicating the probability for unobserved samples
+        ("p0") in cases where no samples with a single count are observed; if
+        this value is not specified, "p0" will default to a Laplace estimation
+        for the current frequency distribution. Please note that this is
+        intended change from the reference implementation by Gale and Sampson.
+
+    Returns
+    -------
+
+    state_prob: dict
+        A dictionary of sample to log-probabilities for all the samples in the
+        frequency distribution.
+
+    unobserved_prob: float
+        The log-probability for samples not found in the frequency
+        distribution.
+    """
+
+    # Try to load the required libraries, failing if they are not available.
+    # These are not included in the global imports of this module in order
+    # to reduce the number of dependencies.
+    from scipy import linalg, stats
```

Comment:
sorry, I have to nitpick one more thing, I was not explicit enough, but the way we handle thirdparties that are only essential for a few methods, we do already in teh beginning of the script (here from algorithm/extra.py)

```
try:
    from sklearn import cluster
except ImportError:
    cluster = False
try:
    import igraph
except ImportError:
    igraph = False
```
And then, in the function:

```
    if not igraph:
        raise ValueError("The package igraph is needed to run this analysis.")
    if not taxa:
        taxa = list(range(1, len(matrix) + 1))
```
Earlier, we even logged a warning, using lingpy's logger, but this should only be on debug level, as people were annoyed by getting a warning all the time they loaded the library.

And since numpy is officially still a dependency, you can just import it as you please in the beginning of the script.

Reply:
Ok for numpy, but I am also using scipy, which is not a dependency.

I will use the solution for third-parties you recommend (and had already recommended, this was a quick fix for later implementing what you had requested, which of course I forgot to do before reopening the PR).